### PR TITLE
#350: Prefix CMake internal variables with `magistrate_` and create a new target named `magistrate` to enable the use of `find_package(magistrate)`

### LINKED ISSUE #350: CMake build variables should be aliased to magistrate

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #350

TODO:
- options should be prefixed with `magistrate`  
- CMake should generate `magistrate` files so it's possible to use `find_package(magistrate)`

### CODE REVIEW

File: ci/build_cpp.sh [hunk 1]
```diff
@@ -41,7 +41,6 @@ cmake -G "${CMAKE_GENERATOR:-Ninja}" \
       -Dcheckpoint_asan_enabled="${CHECKPOINT_ASAN_ENABLED:-0}" \
       -Dcheckpoint_ubsan_enabled="${CHECKPOINT_UBSAN_ENABLED:-0}" \
       -Dcheckpoint_serialization_error_checking_enabled="${CHECKPOINT_SERIALIZATION_ERROR_CHECKING_ENABLED:-$is_debug}" \
-      -DCMAKE_EXPORT_COMPILE_COMMANDS=1 \
```
Comment (JacobDomagala) [inline ci/build_cpp.sh:44]: Duplicate, see few lines above

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -67,57 +67,54 @@ foreach(SUB_DIR ${TOP_LEVEL_SUBDIRS})
...
+  message(STATUS "Magistrate: enabling code coverage generation")
   # Add required flags (GCC & LLVM/Clang)
   target_compile_options(
-    ${CHECKPOINT_LIBRARY} PUBLIC
+    ${MAGISTRATE_LIBRARY} PUBLIC
     -O0        # no optimization
     -g         # generate debug info
     --coverage # sets all required flags
   )
-  if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.13)
```
Comment (JacobDomagala) [inline src/CMakeLists.txt:116]: Right now we require CMake 3.17 (vt requires 3.23, we should also bump the version for magistrate)

Comment (JacobDomagala) [reply src/CMakeLists.txt:116]: https://github.com/DARMA-tasking/magistrate/issues/354 created

File: CMakeLists.txt [hunk 3]
```diff
@@ -46,61 +53,55 @@ if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR AND NOT DEFINED CMAKE_CXX_
...
+if (${magistrate_doxygen_enabled})
   include(cmake/load_doxygen.cmake)
 endif()
 
 # Optionally enable address sanitizer library in build
-if (checkpoint_asan_enabled)
-  if(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" OR
-     CMAKE_CXX_COMPILER_ID STREQUAL "Clang" OR
-     (CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
-       AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "4.8"))
```
Comment (JacobDomagala) [inline CMakeLists.txt:76]: Pretty sure we no longer support gcc 4.8 (?)

Comment (cz4rs) [reply CMakeLists.txt:76]: Looks good. [`vt` requires 8.2](https://github.com/DARMA-tasking/vt/wiki/Build-Requirements#supported-compilers) and AFAIK `magistrate`'s requirements should be aligned.

File: CMakeLists.txt [hunk 4]
```diff
@@ -14,25 +14,32 @@ endif()
 include (CTest)
 enable_testing()
```
Comment (cz4rs) [inline CMakeLists.txt]: Off-topic: should this be enabled only when tests or examples are enabled?

Comment (JacobDomagala) [reply CMakeLists.txt]: Probably, also now that you point that out I realize we have duplicate includes in the same file. We should fix that

Comment (JacobDomagala) [reply CMakeLists.txt]: Fixed

**Comment by JacobDomagala** [comment]
Created https://github.com/DARMA-tasking/vt/pull/2310 that uses this branch to build vt (with `find_package(magistrate)` and `vt::lib::magistrate` as CMake target)

**Review by cz4rs — APPROVED** [review]
Looks good to me.

**Review by lifflander — APPROVED** [review]
Looks good

**Pull request merged**